File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long prev=LONG_MIN;
    bool solve(TreeNode* root)
    {
        if(!root)
        return true;

        bool Left=solve(root->left);
        if(!Left)
        return false;

        if(prev>=root->val)
        return false;

        prev=root->val;

        return solve(root->right);
    }

    bool isValidBST(TreeNode* root) 
    {
        return solve(root);
    }
};

File: 98-validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean validateTree(TreeNode root, long minimum, long maximum){
        if(root == null) return true;
        if(root.val <= minimum || root.val >= maximum) return false;

        return validateTree(root.left, minimum, root.val) && validateTree(root.right, root.val, maximum);
    }

    public boolean isValidBST(TreeNode root) {
        return validateTree(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
}